crates/node/service/src/actors/derivation.rs

```diff
@@ -259,18 +256,35 @@ where
         }
     }
 
+    /// Attempts to process the next payload attributes.
+    ///
+    /// There are a few constraints around stepping on the derivation pipeline.
+    /// - The l2 safe head ([`L2BlockInfo`]) must not be the zero hash.
+    /// - The pipeline must not be stepped on with the same L2 safe head twice.
+    /// - Errors must be bubbled up to the caller.
+    ///
+    /// In order to achieve this, the channel to receive the L2 safe head
+    /// [`L2BlockInfo`] from the engine is *only* marked as _seen_ after payload
+    /// attributes are successfully produced. If the pipeline step errors,
+    /// the same [`L2BlockInfo`] is used again. If the [`L2BlockInfo`] is the
+    /// zero hash, the pipeline is not stepped on.
     async fn process(&mut self, _: Self::InboundEvent) -> Result<(), Self::Error> {
         // Only attempt derivation once the engine finishes syncing.
         if !self.engine_ready {
             trace!(target: "derivation", "Engine not ready, skipping derivation.");
             return Ok(());
         }
 
-        // The L2 Safe Head must be advanced before producing new payload attributes.
-        if self.engine_l2_safe_head.borrow().block_info.number <=
-            self.l2_safe_head.block_info.number
-        {
-            debug!(target: "derivation", engine_safe_head = ?self.engine_l2_safe_head.borrow().block_info.number, l2_safe_head = ?self.l2_safe_head.block_info.number, "L2 safe head unchanged");
+        // If the safe head hasn't changed, don't step on the pipeline.
+        if self.engine_l2_safe_head.has_changed().map_or(false, |v| !v) {
```

Comment:
As is, this check would fail to skip derivation if the channel has closed (meaning the engine actor has shut off.) We should hard error in that case.

Reply:
Good point